### PR TITLE
Align async tool requests with sync execution

### PR DESCRIPTION
**Pull request by guerler**
This PR improves async tool request compatibility with the synchronous tool API by addressing a range of legacy parameter parsing edge cases, including parameter resolution, conditional inference, repeat handling, default initialization, type coercion, and deferred dataset materialization for datasets with `extension="auto"`. Most changes are regression-driven and accompanied by focused unit tests covering the corresponding compatibility behavior.

I also ran a full async vs. synchronous comparison across the IUC tool suite.

Current results:
`regressions=42 pass_only=18 agree_pass=6275 agree_fail=297 missing=0`.

See: https://github.com/guerler/tools-iuc/actions/runs/28927029485.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/tool_util/parameters/case.py [hunk 1]
```diff
@@ -177,6 +183,9 @@ def legacy_from_string(parameter: ToolParameterT, value: Any | None, warnings: l
                     if column_name_value_match:
                         column_part = column_name_value_match.group(1)
                         result_value = int(column_part)
+    elif isinstance(parameter, TextParameterModel) and isinstance(value, (bool, int, float)):
```
Comment (jmchilton) [inline lib/galaxy/tool_util/parameters/case.py]: 🤯 - really - this appears in the wild? This is terrible.

Please add if WARN_ON_UNTYPED_XML_STRINGS I guess and a test tool that exercises this. Do you know what tool in the wild did this?

Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/case.py]: Actually I think we should just break this. That tool is going out of its way to specify a JSON value - that tool has a bug and we shouldn't regress Galaxy for this.

File: lib/galaxy/tool_util_models/parameters.py [hunk 2]
```diff
@@ -326,6 +326,11 @@ def _json_schema_extra_for_validators(validators: Sequence[VT]) -> dict[str, Any
             # Python re.match anchors at start; JSON Schema pattern does not
             if not pattern.startswith("^"):
                 pattern = "^" + pattern
+            # Python ``re`` (used by pydantic's ``pattern``) doesn't recognise POSIX
+            # character classes; defer such patterns to ``statically_validate`` which uses
+            # the ``regex`` module (seurat_plot).
+            if "[:" in pattern and ":]" in pattern:
+                continue
```
Comment (jmchilton) [inline lib/galaxy/tool_util_models/parameters.py:334]: Can you redo this commit with a test tool that exercises this - I'd like to understand it in context.

**Comment by jmchilton** [comment]
Can I have this in smaller chunks - 54fab14b84db6c5ab32257cbee6e8b336be1bdc0 and e7702523c1e8e1969fd12e35844e15214e456c7e look fine, 5a1c9d855bb35740fd23979432b13a9091ffe579 looks great.
I commented on 9756a9c95cfa132781fa2138602b50ccb1194dd5 - I am fine with value="" -> None change there and including Floats. The other change of coercing value_json is wild and I think we shouldn't include it even if it breaks tools. cc989c47d1870019417d6f4feb57306f8a9af89e looks maybe okay - but I want to understand in context and requested a test case.

The other changes are really hard to read - they seem to be mostly emulating older, broken logic so it makes sense that they hard to read - this isn't a critique on this PR but the past behavior. But I think should be in their own PR? Marius went back and forth a lot on the complex logic that is already there and I'd like to take a day to understand these changes.

**Comment by guerler** [comment]
@jmchilton Thanks for review! and yes, some of the legacy compatibility fixes in `case.py` aren't easy to read, and it makes sense to review them critically to determine whether they're actually justified or whether we should accept failure and fix upstream tool declarations instead.

I split the changes into smaller PRs, dropped the `value_json` coercion, added the requested test coverage.

I also prepared an preliminary upstream `tools-iuc` PR that fixes a number of corner cases directly, which might let us remove some of the compatibility logic on the Galaxy side: https://github.com/galaxyproject/tools-iuc/pull/8171

File: lib/galaxy/tool_util/parameters/convert.py [hunk 3]
```diff
@@ -630,9 +630,11 @@ def _initialize_repeat_state(parameter: RepeatParameterModel, tool_state: dict[s
     if parameter_name not in tool_state:
         tool_state[parameter_name] = []
     repeat_instances = cast(list[dict[str, Any]], tool_state[parameter_name])
-    if parameter.min:
-        while len(repeat_instances) < parameter.min:
-            repeat_instances.append({})
+    # Match sync's ``Repeat.get_initial_value()``: seed ``max(min, default)`` empty
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/convert.py]: Questionable change, this should probably be a request that gets rejected.
If you're providing a request it should probably be valid ? An actual failing test as well as the real-life case that prompted this would be necessary for a review.

File: lib/galaxy/tool_util/parameters/convert.py [hunk 4]
```diff
@@ -494,6 +495,11 @@ def _fill_default_for(tool_state: dict[str, Any], parameter: ToolParameterT) ->
     if isinstance(parameter, (IntegerParameterModel, FloatParameterModel, HiddenParameterModel)):
         if parameter_name not in tool_state:
             tool_state[parameter_name] = parameter.value
+    elif isinstance(parameter, ColorParameterModel):
+        # Legacy ``value=""`` on an optional color means "no default"; treat as unset (arriba).
+        if parameter_name not in tool_state:
+            default = parameter.value
+            tool_state[parameter_name] = None if default == "" else default
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/convert.py]: It would seem more correct to assign the appropriate type to the default value. i.e a default value should be a valid color or None. At parse time i would think, probably depending on the tool profile so modern tools are required to set value_json="null"

File: lib/galaxy/tool_util/parameters/convert.py [hunk 5]
```diff
@@ -630,9 +636,11 @@ def _initialize_repeat_state(parameter: RepeatParameterModel, tool_state: dict[s
     if parameter_name not in tool_state:
         tool_state[parameter_name] = []
     repeat_instances = cast(list[dict[str, Any]], tool_state[parameter_name])
-    if parameter.min:
-        while len(repeat_instances) < parameter.min:
-            repeat_instances.append({})
+    # Match sync's ``Repeat.get_initial_value()``: seed ``max(min, default)`` empty
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/convert.py]: That comment will not age well once we reduce basic.py, if we need the comment you'd probably want to say legacy ?

File: lib/galaxy/model/deferred.py [hunk 6]
```diff
@@ -256,6 +265,20 @@ def ensure_materialized(
         self._previously_materialized[dataset_instance.id] = materialized_dataset_instance
         return materialized_dataset_instance
 
+    def _sniff_deferred_extension(
+        self,
+        path: str,
+        dataset_instance: HistoryDatasetAssociation | LibraryDatasetDatasetAssociation,
+    ) -> None:
+        """Replace ``extension="auto"`` with a sniffed extension after download."""
```
Comment (mvdbeek) [inline lib/galaxy/model/deferred.py]: after materialization ?

File: test/unit/data/test_dataset_materialization.py [hunk 7]
```diff
@@ -434,6 +434,48 @@ def test_deferred_hdas_with_deferred_metadata():
...
+    materialized_hda = materializer.ensure_materialized(deferred_hda)
+
+    assert materialized_hda.dataset is not None
+    assert materialized_hda.dataset.state == "ok"
+    # The fixture source is a .bed file; content sniffing must upgrade "auto" to "bed".
+    assert materialized_hda.extension == "bed"
+
+
+def test_deferred_hda_without_registry_keeps_auto_extension():
+    # Without a datatypes_registry the sniff step is a no-op; extension stays "auto".
```
Comment (mvdbeek) [inline test/unit/data/test_dataset_materialization.py]: In what case do we need this variant ?

File: lib/galaxy/tool_util/parameters/case.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/tool_util/parameters/case.py]: I guess to follow up, which tool is this ?

File: test/unit/data/test_dataset_materialization.py (hunk 7 above)
Comment (guerler) [reply test/unit/data/test_dataset_materialization.py]: The registry is optional and sniffing only happens when it's passed. This test checks the other case: with no registry, the extension stays "auto".

File: lib/galaxy/tool_util/parameters/convert.py (hunk 4 above)
Comment (guerler) [reply lib/galaxy/tool_util/parameters/convert.py]: Moved to parse time: an optional color with value="" is now built as value=None, so the model always holds a valid color or None.

File: lib/galaxy/tool_util/parameters/convert.py (hunk 5 above)
Comment (guerler) [reply lib/galaxy/tool_util/parameters/convert.py]: Done.

File: lib/galaxy/tool_util/parameters/convert.py (hunk 3 above)
Comment (guerler) [reply lib/galaxy/tool_util/parameters/convert.py]: This coercion is gone from convert.py. The empty color value is now turned into None at parse time, so bad requests still fail validation. Covered by gx_color_optional_no_default in test_parameter_convert.py.

File: lib/galaxy/model/deferred.py (hunk 6 above)
Comment (guerler) [reply lib/galaxy/model/deferred.py]: Yes, it runs during materialization, on the downloaded file before it's saved to the object store, which is the first point the data exists locally. I updated the comment accordingly.

File: lib/galaxy/tool_util/parameters/case.py (hunk 1 above)
Comment (guerler) [reply lib/galaxy/tool_util/parameters/case.py]: It's `map_param_value` (IUC). Its `to` field is `type="text"`, but a few tests pass `value_json="true"`/`"false"` to it. That's the tool passing a JSON boolean to a text parameter, so per @jmchilton's feedback I removed the coercion and left it to be fixed in the tool instead of the framework.

File: test/unit/data/test_dataset_materialization.py (hunk 7 above)
Comment (mvdbeek) [reply test/unit/data/test_dataset_materialization.py]: but that's not a useful behavior, so when is this needed ? make it a requirement ?

Comment (guerler) [reply test/unit/data/test_dataset_materialization.py]: I dropped the test. The case this covers are async tool tests that pass an input by URL without explicit ftype. Those default to `auto` and get materialized in `queue_jobs`, and without sniffing the tool runs on an untyped `auto` dataset, while the sync path sniffs during `fetch`. The tool execution paths already always pass the registry, so sniffing isn't optional so there is no test needed as you pointed out.

File: lib/galaxy/model/deferred.py [hunk 8]
```diff
@@ -81,6 +83,7 @@ def __init__(
         file_sources: ConfiguredFileSources | None = None,
         sa_session: Session | None = None,
         user_context: OptionalUserContext = None,
+        datatypes_registry: Registry | None = None,
```
Comment (mvdbeek) [inline lib/galaxy/model/deferred.py]: What i mean is why would you allow None at all ? xref https://github.com/galaxyproject/galaxy/pull/23084/changes/BASE..d9b0e38c19e2f0994845dd330ca2fa17778e2a7b#r3576622405. if you change that don't forget to update the comments below.

Comment (guerler) [reply lib/galaxy/model/deferred.py]: `None` is not allowed anymore. Thx.

File: lib/galaxy/tool_util/parameters/convert.py (hunk 3 above)
Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/convert.py]: @guerler This comment is not aimed at the right change I think. You mentioned color stuff but this is about the repeat handling? 

Claude and I went in circles about the repeat thing and I still don't understand. I will try to come back to this and figure out where I stand but in the mean time I think a test case would be great (if not just extracting this out into a clean PR with just the repeat changes). 

The latest Claude comment refinement that just is so verbose 😿 is below. 

<details>

Picking up the `_initialize_repeat_state` thread (https://github.com/galaxyproject/galaxy/pull/23084#discussion_r3576547722) — the reply there covers the color coercion, so the `max(min, default)` question is still open and the code is still in the diff.

The absent-repeat case and the explicit-`[]` case behave differently and should be settled separately. The first matches production sync. The second is where the objection lands, and this PR introduces a new instance of it.

## The absent case matches the sync reference

The behavioral reference here is `_populate_state_legacy`. `services/tools.py:378` defaults `input_format` to `"legacy"`, and the client never sends it — zero references anywhere under `client/`, including the generated schema. (`client/src/components/Tool/submit/index.js` routes to `/api/jobs` when `enable_tool_requests && enable_celery_tasks && has_parameters`; those defaults were flipped off in #23113 for the release, so this is the path being switched back on, not a hypothetical one.)

```python
# lib/galaxy/tools/parameters/__init__.py:580
            del group_state[:]
            while True:
                rep_prefix = f"{key}_{rep_index}"
                rep_min_default = input.default if input.default > input.min else input.min
                if (
                    not any(incoming_key.startswith(rep_prefix) for incoming_key in incoming.keys())
                    and rep_index >= rep_min_default
                ):
                    break
```

With no incoming keys for the repeat, the break is suppressed until `rep_index >= max(min, default)`, so that many instances are appended and recursively filled. A `<repeat default="1">` submitted with nothing gets one instance under sync and got zero under the async request path before this PR — a real divergence, in scope for "Align async tool requests with sync execution".

Filling absent repeats from the tool's declaration is already the established async contract for `min`:

```yaml
# test/unit/tool_util/parameter_specification.yml:2108
gx_repeat_boolean_min:
  request_valid:
    # even though a minimum is set here - each instance of the repeat does not require a value
    # so we can skip the repeat all together and Galaxy I think will just use the defaults at
    # each stage - see API tests -> test_tool_execute.py -> test_optional_repeats_with_mins_filled_id
    - {}
```

`test_optional_repeats_with_mins_filled_id` doesn't take the `tool_input_format` fixture, so it pins the legacy side only, not the request side.

One counter-argument worth resolving explicitly rather than by omission: `min` is a constraint the request must satisfy, while `default` is closer to a form-presentation hint — it is what `Repeat.get_initial_value()` feeds into `populate_model`'s form cache. A structured request is not a form. That doesn't outweigh the legacy behavior above, but "`min` and `default` are the same kind of thing" is an assumption, not a given.

## The explicit-`[]` case, including one this PR adds

What is already rejected: `strict` defaults to `True` (`services/jobs.py:81`) and `strictify` only runs under `if not strict:` (`:255`), documented as "drops support for some inconsistent legacy behavior". On the default path, `parameter: []` against a `min="2"` repeat is rejected by the request model — `min_length` on the generated `RootModel` and `parameter_specification.yml:2119` agree.

The live site is `_fill_default_for` (`convert.py:536`), reached through `fill_static_defaults`, which has no opt-out: `tools/__init__.py:2188-2189` for every async tool execution, `workflow/modules.py:2402` for every workflow step.

```python
    # Seed ``max(min, default)`` empty instances when the request supplies none.
    floor = max(parameter.min or 0, parameter.default or 0)
    while len(repeat_instances) < floor:
        repeat_instances.append({})
```

The comment says "when the request supplies none"; the code does not check that. Before this PR `floor` was `min`, so a `min="0"` repeat was never padded. After it, `min="0" default="2"` with an explicit `parameter: []` passes request validation (`min_length` 0) and is then padded to two instances at job-internal time.

With tool requests switched back on, that is reachable from the tool form as a user-visible regression:

1. Tool declares `<repeat default="2">`. `populate_state` seeds `group_state` with two instances via `Repeat.get_initial_value()`, and `populate_model.py:36` ships them to the form as `tool_dict["cache"]`.
2. The form renders two blocks. The user deletes both.
3. `buildNestedState` (`client/src/components/Form/utilities.js:180`) emits `result[nodeName] = items` unconditionally, so the request carries `rep: []` — pinned by `utilities.test.js:619`.
4. `fill_static_defaults` pads it back to two.

The user gets back the repeat instances they explicitly deleted. The request is valid and says zero; we override it. Legacy sync cannot express that distinction — flat keys have no spelling for "zero repeats", which is why `rep_min_default` exists in that loop — but the structured format can, and the form already relies on it.

Also worth noting from the same trace: because the form receives the `default` instances materialized in `cache` and echoes them back explicitly, the repeat key is *always* present in a tool-form request. The `default` branch of `_initialize_repeat_state` therefore never fires for form submissions — it only affects API callers that omit the key entirely, and workflow steps. Whatever parity argument motivates `default` support, it does not come from the tool form.

The padding is not only cosmetic either. Each seeded instance gets its inner defaults filled, so for a repeat containing a required child (a `data` param), adding `default="1"` converts a previously-valid empty repeat into a hard `job_internal` validation failure.

Separating the constraint from the seed:

```python
    supplied = parameter_name in tool_state
    if not supplied:
        tool_state[parameter_name] = []
    repeat_instances = cast(list[dict[str, Any]], tool_state[parameter_name])
    # ``min`` is a constraint - always pad up to it. ``default`` only seeds a request that
    # omitted the repeat entirely; an explicit [] means the caller asked for zero.
    floor = parameter.min or 0
    if not supplied:
        floor = max(floor, parameter.default or 0)
    while len(repeat_instances) < floor:
        repeat_instances.append({})
```

Both conditions are load-bearing. Gating the whole floor on absence — or on `repeat_instances` being empty — breaks `test_parameter_convert.py:324`, which asserts `{"parameter": [{}]}` against `gx_repeat_boolean_min` (`min="2"`) pads to two; `[{}]` is both present and non-empty, so either gate leaves it at one and fails `min_length=2`. Gating on emptiness additionally fixes nothing, since an explicit `[]` is empty and gets seeded anyway.

Resulting behavior:

| declaration | request | instances |
|---|---|---|
| `min="0" default="2"` | absent | 2 |
| `min="0" default="2"` | `[]` | **0** (currently 2) |
| `min="0" default="2"` | `[{}]` | 1 |
| `min="2"` | absent | 2 |
| `min="2"` | `[]` | 2 |
| `min="2"` | `[{}]` | 2 |

Green on `test_parameter_convert.py`, `test_parameter_specification.py`, `test_parameter_test_cases.py` (46 passed). The rest of `test/unit/tool_util` was not run.

## Possibly cleaner: clamp in `factory.py`

Sync clamps `default` into `[min, max]` at parse time:

```python
# lib/galaxy/tools/__init__.py:1860
                # Force default to be within min-max range
                group_r.default = cast(int, min(max(group_r.default, group_r.min), group_r.max))
```

which is why `Repeat.get_initial_value()`'s bare `range(self.default)` already is `max(min, default)` in practice, and why `rep_min_default` at `:583` is belt-and-braces. `factory.py:397` does not replicate the clamp — it is `int(default_raw) if default_raw is not None else 0`.

Clamping there would give `RepeatParameterModel.default` the same invariant as `Repeat.default`, let `_initialize_repeat_state` use `parameter.default` directly, and fix a divergence the current version has: `<repeat default="5" max="2">` seeds 5 async vs. 2 sync.

Separately, `fill_static_defaults` takes `profile` and threads it nowhere into `_fill_defaults`. Legacy-emulation behaviors here are usually profile-gated; this one arrives ungated for all profiles.

## "Align with sync" is underdetermined — legacy and 21.01 disagree

`populate_state(input_format="21.01")` seeds `group_state` from `get_initial_value()` and then:

```python
# lib/galaxy/tools/parameters/__init__.py:460
            repeat_incoming = incoming.get(input_name) or []
            if repeat_incoming and (len(repeat_incoming) > input.max or len(repeat_incoming) < input.min):
                errors[input_name] = "The number of repeat elements is outside the range specified by the tool."
            else:
                del group_state[:]
```

Empty → guard short-circuits → `else` → `del group_state[:]` discards the seed. So 21.01 honors neither `min` nor `default` on an absent repeat, and does not validate `min` either. `populate_state_async` (`:692`) differs again: it indexes `incoming[input_name]` directly, so the key must exist — that is the load-bearing half of `_initialize_repeat_state`, independent of any floor — and its `min` check has no short-circuit, so it does enforce `min`.

Legacy is the right reference; 21.01's missing `min` check on an absent repeat reads as an oversight rather than a design position and probably deserves its own issue. Worth a sentence in the commit message rather than leaving it implied by "match sync".

## Where does this reproduce?

`da4814c4eb` named the case, and it was dropped in the reword:

```python
# Match sync's ``Repeat.get_initial_value()``: seed ``max(min, default)`` empty
# instances when the request supplies none (bigscape).
```

Worth restoring, ideally in the commit message with the concrete symptom.

`da4814c4eb` is titled *"Honor `<repeat default="N">` in test-case merging"*, but `_initialize_repeat_state` is called from `_strictify_parameter` (`convert.py:201`), `_fill_default_for` (`:536`) and `_fill_url_default_for` (`:594`) — all request/execution paths. Test-case merging does its own padding in `case.py:435` and never touches this helper:

```python
        if tool_input.min is not None:
            while len(repeat_instance_inputs) < tool_input.min:
                repeat_instance_inputs.append([])
```

Per the `cache` trace above, the tool form cannot be the source: it always sends the repeat key explicitly, so the `default` branch never runs for it. That points at the planemo/test-case path — matching the commit title — or a saved workflow step state going through `modules.py:2402`. If it is test-case merging, `case.py` is the right home and the execution path does not need to change.

## Test coverage

No tool anywhere in the tree uses `<repeat default=`, and no test constructs `RepeatParameterModel(default=N)`, so the field is entirely uncovered. Assets to reuse:

1. `test/functional/tools/parameters/gx_repeat_boolean_default.xml`, mirroring `gx_repeat_boolean_min.xml`.
2. A `gx_repeat_boolean_default:` entry in `parameter_specification.yml` pinning absent vs. explicit `[]` vs. too-few, across `request` and `job_internal`. Include a `min="0" default="1"` case around a required child input — that is where seeding turns a valid empty repeat into a hard failure.
3. An API test in `test_tool_execute.py` on the existing fixture:

```python
# lib/galaxy_test/api/conftest.py:154
@pytest.fixture(params=["legacy", "21.01", "request"])
def tool_input_format(request) -> Iterator[DescribeToolInputs]:
```

That last one runs the same submission through legacy, 21.01 and the async request path, which settles the question empirically. On the reading above it should show legacy and 21.01 already disagreeing on a `default="2"` repeat, making the choice explicit rather than inherited.
</details>

Comment (guerler) [reply lib/galaxy/tool_util/parameters/convert.py]: Ok, this discussion deserves a separate follow-up PR. I removed those changes from this PR and will open a separate one.

**Comment by jmchilton** [comment]
Thanks so much for breaking this down for me - I'm really up to review the parts you extracted also if you want to open them in isolation. I really appreciate all the fixes and your flexibility here.
